### ISSUE TITLE
Nan value for loss with keras resnet50 model

### ISSUE THREAD

**Issue by arman-kazemi** [tensorflow/tpu#278]
I am trying to train the ResNet50 model with Keras in the experimental models, but the loss value is a Nan. I have slightly modified the code in the "main" function to run with GPUs:
```
def main(unused_argv):
  assert FLAGS.data is not None, 'Provide training data path via --data.'

  batch_size = FLAGS.num_cores * PER_CORE_BATCH_SIZE
  training_steps_per_epoch = int(APPROX_IMAGENET_TRAINING_IMAGES / batch_size)
  validation_steps = int(IMAGENET_VALIDATION_IMAGES // batch_size)

  model_dir = FLAGS.model_dir if FLAGS.model_dir else DEFAULT_MODEL_DIR
  logging.info('Saving tensorboard summaries at %s', model_dir)
  logging.info('Use global batch size: %s.', batch_size)
  logging.info('Training model using data in directory "%s".', FLAGS.data)

#  with strategy.scope():
  logging.info('Building Keras ResNet-50 model')
  model = resnet_model.ResNet50(num_classes=NUM_CLASSES)

  logging.info('Compiling model.')
  metrics = ['sparse_categorical_accuracy']

  model.compile(
        optimizer=SGD(
            lr=BASE_LEARNING_RATE, momentum=0.9, nesterov=True),
        loss='sparse_categorical_crossentropy',
        metrics=metrics)

  imagenet_train = imagenet_input.ImageNetInput(
      is_training=True, data_dir=FLAGS.data, batch_size=batch_size,
      use_bfloat16=USE_BFLOAT16)
  imagenet_eval = imagenet_input.ImageNetInput(
      is_training=False, data_dir=FLAGS.data, batch_size=batch_size,
      use_bfloat16=USE_BFLOAT16)

  lr_schedule_cb = LearningRateBatchScheduler(
      schedule=learning_rate_schedule_wrapper(training_steps_per_epoch))
  tensorboard_cb = eval_utils.TensorBoardWithValidation(
      log_dir=model_dir,
      validation_imagenet_input=imagenet_eval,
      validation_steps=validation_steps,
      validation_epochs=[30, 60, 90])

  training_callbacks = [lr_schedule_cb, tensorboard_cb]

  model.fit(
      imagenet_train.input_fn(),
      epochs=EPOCHS,
      steps_per_epoch=training_steps_per_epoch,
      callbacks=training_callbacks)

  model_saving_utils.save_model(model, model_dir, WEIGHTS_TXT)

```
I use a global batch size of 64 with 1 GPU. TensorFlow version is 1.12.

I see this comment in the code:
TODO(b/120613626): Use default constructor once the bug generating Nan loss is fixed.
Can you tell me how to fix this problem?

**Comment by arman-kazemi** [comment]
The problem is in the imagenet_input.py file in the "dataset_parser" function. The labels from the training data are the name of the class synset, but the labels for the validation data which comes from the validation.txt file are integers in the range [0,1000). Once the dataset_parser function reads the labels from the tensor slices, it assumes that validation and training labels are coherent which is not the case. The labels are in the range [0,2000). From [0,1000) is for validation data and from [1000,2000) is for training data. For sparse_categorical_crossenteropy, the labels need to be in integer format as well.
I fixed the problem by simply subtracting 1000 from the training labels:
```
    # Subtract one so that labels are in [0, 1000), and cast to float32 for
    # Keras model.
    if self.is_training:
       label = tf.cast(tf.cast(
           tf.reshape(parsed['image/class/label'], shape=[1]), dtype=tf.int32) - 1001,
                    dtype=tf.int32)#tf.float32)
    else:
       label = tf.cast(tf.cast(
           tf.reshape(parsed['image/class/label'], shape=[1]), dtype=tf.int32) - 1,
                    dtype=tf.int32)#tf.float32)
```
